src/java/org/apache/cassandra/index/sai/disk/v1/V1OnDiskFormat.java

```diff
@@ -0,0 +1,167 @@
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.compaction.OperationType;
+import org.apache.cassandra.db.lifecycle.LifecycleNewTracker;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.StorageAttachedIndex;
+import org.apache.cassandra.index.sai.disk.ColumnIndexWriter;
+import org.apache.cassandra.index.sai.disk.PerSSTableComponentsWriter;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.format.OnDiskFormat;
+import org.apache.cassandra.index.sai.disk.v1.writers.MemtableIndexWriter;
+import org.apache.cassandra.index.sai.disk.v1.writers.SSTableComponentsWriter;
+import org.apache.cassandra.index.sai.disk.v1.writers.SSTableIndexWriter;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.NamedMemoryLimiter;
+import org.apache.cassandra.index.sai.utils.SAICodecUtils;
+import org.apache.cassandra.schema.CompressionParams;
+import org.apache.lucene.store.IndexInput;
+
+import static org.apache.cassandra.index.sai.StorageAttachedIndex.SEGMENT_BUILD_MEMORY_LIMITER;
+import static org.apache.cassandra.utils.FBUtilities.prettyPrintMemory;
+
+public class V1OnDiskFormat implements OnDiskFormat
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    private static final EnumSet<IndexComponent.Type> LITERAL_COMPONENT_TYPES = EnumSet.of(IndexComponent.Type.COLUMN_COMPLETION_MARKER,
+                                                                                           IndexComponent.Type.META,
+                                                                                           IndexComponent.Type.TERMS_DATA,
+                                                                                           IndexComponent.Type.POSTING_LISTS);
+    private static final EnumSet<IndexComponent.Type> NUMERIC_COMPONENT_TYPES = EnumSet.of(IndexComponent.Type.COLUMN_COMPLETION_MARKER,
+                                                                                           IndexComponent.Type.META,
+                                                                                           IndexComponent.Type.KD_TREE,
+                                                                                           IndexComponent.Type.KD_TREE_POSTING_LISTS);
+
+    public static final V1OnDiskFormat instance = new V1OnDiskFormat();
+
+    private V1OnDiskFormat()
+    {}
+
+    @Override
+    public boolean isGroupIndexComplete(IndexDescriptor indexDescriptor)
+    {
+        return indexDescriptor.registerSSTable().hasComponent(IndexComponent.GROUP_COMPLETION_MARKER);
+    }
+
+    @Override
+    public boolean isColumnIndexComplete(IndexDescriptor indexDescriptor, IndexContext indexContext)
```

Comment:
Similar problem here - the method naming suggests is is side-effect free, but actually it is not.